### ISSUE TITLE
DoNotMock false negatives

### ISSUE THREAD

**Issue by ZacSweers** [slackhq/slack-lints#53]
We have a case like this in the slack codebase that shows as red in the IDE but does not result as an error in the lint command line invocation

```kotlin
class SomeTest : SlackTest() {
    @Mock lateinit var reaction: UserReacted // UserReacted is a data class
}
```

It doesn't appear to be disabled as it's visible in the list of `ExtraIssues`

<img width="817" alt="image" src="https://github.com/slackhq/slack-lints/assets/1361086/1f892814-336c-41b1-8049-bc1c9b2e48d4">

**Comment by ZacSweers** [comment]
Wondering if it's to do with our older minApi and hoping bumping to latest lint will fix it

**Comment by ZacSweers** [comment]
Big sigh https://issuetracker.google.com/issues/283654244

Some potential workarounds
A. Parse metadata ourselves by shading in and using kotlinx-metadata. Worried about overhead and caching though.
B. See if this is fixed by `checkDependencies=true`, but that comes with its own major costs.
C. Write a compiler plugin that _adds_ `@DoNotMock` annotations to these implicit DoNotMock types, then rely on _that_ lint detector to recognize them.

I think A or C are the best options. C solves the DoNotMock case only, while A solves it more generally (and probably other detectors with similar issues). If we can make sure it's performant, A seems like the more ideal solution.
